Userland/Shell/Shell.h

```diff
@@ -93,8 +93,52 @@ class Shell : public Core::Object {
         Optional<AST::Position> position;
     };
 
+    struct RunnablePath {
+        enum class Kind {
+            Builtin,
+            Function,
+            Alias,
+            Executable,
+        };
+
+        Kind kind;
+        String path;
+
+        bool operator<(RunnablePath const& other) const
+        {
+            return path < other.path;
+        }
+
+        bool operator==(RunnablePath const&) const = default;
+    };
+
+    struct RunnablePathComparator {
+        int operator()(RunnablePath const& lhs, RunnablePath const& rhs)
+        {
+            if (lhs.path > rhs.path)
+                return 1;
+
+            if (lhs.path < rhs.path)
+                return -1;
+
+            return 0;
+        }
+
+        int operator()(StringView lhs, RunnablePath const& rhs)
+        {
+            if (lhs > rhs.path)
+                return 1;
+
+            if (lhs < rhs.path)
+                return -1;
+
+            return 0;
+        }
+    };
+
     int run_command(StringView, Optional<SourcePosition> = {});
-    bool is_runnable(StringView);
+    Optional<RunnablePath> get_runnable_path(StringView);
+    Optional<String> get_help_path(Vector<RunnablePath> visited, RunnablePath const& runnable_path);
```

Comment:
Drop the `get_`, how about `runnable_path()` and `help_path_for()`?

Reply:
Done :) I gave `runnable_path_for()` the `_for` prefix as well to avoid some naming similarity with locals in a few other methods.